src/Compilers/Core/Portable/Syntax/SyntaxNode.cs

```diff
@@ -1649,5 +1675,179 @@ private bool HasErrorsSlow()
 
             return clone;
         }
+
+        /// <summary>
+        /// Gets the index of the given <paramref name="targetNode"/> in the current SyntaxNode.
+        /// </summary>
+        /// <returns>
+        /// <c>-1</c> if <paramref name="targetNode"/> isn't a direct child.
```

Comment:
is this an expected case.  who calls this if it's not a direct child?

Reply:
It's a case supported for resiliency, allowing edge-cases like an invalid token or node which _claims_ to have a specific parent even, and to allow this method to be used in other scenarios.

Reply:
> allowing edge-cases like an invalid token or node which claims to have a specific parent even

are tehre tests that hit this that can demonstrate when this occurrs.  I don't know what it means to have an 'invalid token or not which claims to have a specific parent'.  The roslyn syntax model is built on top of very specific invariants.  Either we have those trees or we don't.  If not, we shouldn't be resilient to things that cannot be occur.  it just makes things much more confusing to reason about.

Reply:
This method doesn't assume anything about its caller, plus the alternatives of iterating forever or throwing an exception, instead of simply returning -1 (like when most `IndexOf` methods can't find the requested entity) would be _more_ unexpected than this method's current return behavior.

Reply:
i don't understand.  this is an internal method.  as such, it will only be called by us.  we will then absolutely know who calls it, and with what values, and we can specify exactly the contract we have.

>would be more unexpected than this method's current return behavior.

that is not at al true.  we throw exceptions *all the time* when a caller does soemthing bad/unexpected.  That's how we fix our callers to be correct.

Reply:
For a method whose general purpose is to search for something in a list and return the index of that item,
returning -1 if that item can't be found is the expected return behavior, throwing an error when that item can't be found here would be an anti-pattern

Reply:
This is not a general purpose, or public, method.  it is an internal method specific to our own algorithms, which should operate under the invariants we have about our trees.  This is how we do things in roslyn.  If something should be impossible for the caller to do, do not try to gracefully be resilient to it.  Instead, check and validate that the caller is not screwing things up.  If htey are, the code should make that clear immediately so we fix this up.

> throwing an error when that item can't be found here would be an anti-pattern

No.  It is hte correct thing to do when we own all callsites and know that this must be called on our own trees that have follow our invariants.  

If you want to discuss this further we can take it to discord.gg/csharp.  You can reach me under the 'metasyntactic' handle.